cvmfs/catalog_traversal_parallel.h

```diff
@@ -0,0 +1,457 @@
+                        const shash::Any &hash,
+                        const unsigned tree_level,
+                        const unsigned history_depth,
+                        CatalogTN *parent = NULL) :
+      CatalogTraversal<ObjectFetcherT>::CatalogJob(path, hash, tree_level,
+                                                   history_depth, parent) {
+      atomic_init32(&children_unprocessed);
+    }
+
+    void WakeParents() {
+      this->NotifyListeners(0);
+    }
+
+    atomic_int32 children_unprocessed;
+  };
+
+ public:
+  /**
+   * Starts the traversal process.
+   * After calling this methods CatalogTraversal will go through all catalogs
+   * and call the registered callback methods for each found catalog.
+   * If something goes wrong in the process, the traversal will be cancelled.
+   *
+   * @return       true, when all catalogs were successfully processed. On
+   *               failure the traversal is cancelled and false is returned.
+   */
+  bool Traverse(const TraversalType type = kBreadthFirstTraversal) {
+    const shash::Any root_catalog_hash = this->GetRepositoryRootCatalogHash();
+    if (root_catalog_hash.IsNull()) {
+      return false;
+    }
+    return Traverse(root_catalog_hash, type);
+  }
+
+  /**
+   * Starts the traversal process at the catalog pointed to by the given hash
+   *
+   * @param root_catalog_hash  the entry point into the catalog traversal
+   * @return                   true when catalogs were successfully traversed
+   */
+  bool Traverse(const shash::Any &root_catalog_hash,
+                const TraversalType type = kBreadthFirstTraversal) {
+    // add the root catalog of the repository as the first element on the job
+    // stack
+    if (this->no_repeat_history_ &&
+        catalogs_done_.Contains(root_catalog_hash)) {
+      return true;
+    }
+    effective_traversal_type_ = type;
+    CatalogJob *root_job = new CatalogJob("", root_catalog_hash, 0, 0);
+    PushJob(root_job);
+    return DoTraverse();
+  }
+
+/**
+ * Start the traversal process from a list of root catalogs. Same as 
+ * TraverseRevision function, TraverseList does not traverse into predecessor
+ * catalog revisions and ignores TraversalParameter settings.
+ */
+  bool TraverseList(const HashList &root_catalog_list,
+                    const TraversalType type = kBreadthFirstTraversal) {
+    // Push in reverse order for CatalogTraversal-like behavior
+    HashList::const_reverse_iterator i = root_catalog_list.rbegin();
+    const HashList::const_reverse_iterator iend = root_catalog_list.rend();
+    bool has_pushed = false;
+    {
+      MutexLockGuard m(&catalogs_lock_);
+      for (; i != iend; ++i) {
+        if (this->no_repeat_history_ && catalogs_done_.Contains(*i)) {
+          continue;
+        }
+
+        CatalogJob *root_job = new CatalogJob("", *i, 0, 0);
+        PushJobUnlocked(root_job);
+        has_pushed = true;
+      }
+    }
+    // noop: no catalogs to traverse
+    if (!has_pushed) {
```

Comment:
Is `TraverseList` always called with an initially empty job list? In this case, we would not need the `has_pushed` variable but we could check whether the job list is empty.

Reply:
There is a different problem here: the list can be non-empty, but only filled with root catalogs which were already traversed in the past. All of them will be skipped by the 'continue' on line 132. Even with non-empty list, you'll end up with an empty queue.